File: sdk/typescript/src/github.ts
import { createClient, type Client, type Transport } from "@connectrpc/connect";
import { create } from "@bufbuild/protobuf";
import {
  ExchangeOAuthCodeRequestSchema,
  GetOAuthAuthorizeUrlRequestSchema,
  GitHubService,
} from "@stigmer/protos/ai/stigmer/platform/github/v1/service_pb";
import { wrapError } from "./gen/errors.js";

/** Parameters for getting the OAuth authorize URL. */
export interface GetOAuthAuthorizeUrlParams {
  readonly redirectUri: string;
}

/** Response containing the OAuth authorize URL and state. */
export interface OAuthAuthorizeUrlResponse {
  readonly authorizeUrl: string;
  readonly state: string;
}

/** Parameters for exchanging an OAuth authorization code. */
export interface ExchangeOAuthCodeParams {
  readonly code: string;
  readonly state: string;
  readonly redirectUri: string;
}

/** Response containing the exchanged access token. */
export interface OAuthTokenResponse {
  readonly accessToken: string;
  readonly tokenType: string;
  readonly scope: string;
}

/**
 * Client for GitHub OAuth integration.
 *
 * Provides methods to initiate the OAuth flow (get authorize URL) and
 * exchange the authorization code for an access token. The access token
 * is returned to the caller — the backend never persists it.
 */
export class GitHubClient {
  private readonly github: Client<typeof GitHubService>;

  constructor(transport: Transport) {
    this.github = createClient(GitHubService, transport);
  }

  /** Get the GitHub OAuth authorize URL to redirect the user to. */
  async getOAuthAuthorizeUrl(
    params: GetOAuthAuthorizeUrlParams,
  ): Promise<OAuthAuthorizeUrlResponse> {
    try {
      const resp = await this.github.getOAuthAuthorizeUrl(
        create(GetOAuthAuthorizeUrlRequestSchema, {
          redirectUri: params.redirectUri,
        }),
      );
      return {
        authorizeUrl: resp.authorizeUrl,
        state: resp.state,
      };
    } catch (e) {
      throw wrapError(e);
    }
  }

  /** Exchange an OAuth authorization code for an access token. */
  async exchangeOAuthCode(
    params: ExchangeOAuthCodeParams,
  ): Promise<OAuthTokenResponse> {
    try {
      const resp = await this.github.exchangeOAuthCode(
        create(ExchangeOAuthCodeRequestSchema, {
          code: params.code,
          state: params.state,
          redirectUri: params.redirectUri,
        }),
      );
      return {
        accessToken: resp.accessToken,
        tokenType: resp.tokenType,
        scope: resp.scope,
      };
    } catch (e) {
      throw wrapError(e);
    }
  }
}

File: sdk/go/github.go
package stigmer

import (
	"context"

	"github.com/stigmer/stigmer/sdk/go/v3/internal/gen"
	githubv1 "github.com/stigmer/stigmer/sdk/go/v3/proto/ai/stigmer/platform/github/v1"
	"google.golang.org/grpc"
)

// GetOAuthAuthorizeUrlParams configures a request to get the GitHub OAuth authorize URL.
type GetOAuthAuthorizeUrlParams struct {
	RedirectURI string
}

// OAuthAuthorizeUrlResponse holds the authorize URL and CSRF state returned by the platform.
type OAuthAuthorizeUrlResponse struct {
	AuthorizeURL string
	State        string
}

// ExchangeOAuthCodeParams configures a request to exchange a GitHub OAuth authorization code.
type ExchangeOAuthCodeParams struct {
	Code        string
	State       string
	RedirectURI string
}

// OAuthTokenResponse holds the access token returned by the GitHub OAuth exchange.
type OAuthTokenResponse struct {
	AccessToken string
	TokenType   string
	Scope       string
}

// GitHubClient provides GitHub OAuth integration against the Stigmer platform.
type GitHubClient struct {
	github githubv1.GitHubServiceClient
}

func newGitHubClient(conn grpc.ClientConnInterface) *GitHubClient {
	return &GitHubClient{github: githubv1.NewGitHubServiceClient(conn)}
}

// GetOAuthAuthorizeUrl returns the GitHub OAuth authorize URL to redirect the user to.
func (g *GitHubClient) GetOAuthAuthorizeUrl(ctx context.Context, params *GetOAuthAuthorizeUrlParams) (*OAuthAuthorizeUrlResponse, error) {
	resp, err := g.github.GetOAuthAuthorizeUrl(ctx, &githubv1.GetOAuthAuthorizeUrlRequest{
		RedirectUri: params.RedirectURI,
	})
	if err != nil {
		return nil, gen.WrapErr(err)
	}
	return &OAuthAuthorizeUrlResponse{
		AuthorizeURL: resp.GetAuthorizeUrl(),
		State:        resp.GetState(),
	}, nil
}

// ExchangeOAuthCode exchanges an OAuth authorization code for an access token.
func (g *GitHubClient) ExchangeOAuthCode(ctx context.Context, params *ExchangeOAuthCodeParams) (*OAuthTokenResponse, error) {
	resp, err := g.github.ExchangeOAuthCode(ctx, &githubv1.ExchangeOAuthCodeRequest{
		Code:        params.Code,
		State:       params.State,
		RedirectUri: params.RedirectURI,
	})
	if err != nil {
		return nil, gen.WrapErr(err)
	}
	return &OAuthTokenResponse{
		AccessToken: resp.GetAccessToken(),
		TokenType:   resp.GetTokenType(),
		Scope:       resp.GetScope(),
	}, nil
}
